### ISSUE TITLE
feat(npu): Zaya whole-layer per-ctx ELF decode (M3/M4) — author whole-layer kernels + RuntimeLayerEngine-for-Zaya (the #1776 endgame path)

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#2080]
## Context

Tracking issue for the actual #1776 endgame on the standalone Zaya path: replace the per-op m8 kernel decode (~7.8 ms/layer, 156 ms/tok) with **whole-layer per-ctx ELFs + one `xrt::runlist` per token** — the substrate that already gives the Qwen3-0.6B npu-infer runtime ~0.96 ms/layer (~37 tok/s). This is a port/authoring milestone, not plumbing.

## Why capture is impossible (authoring required)

The Qwen per-ctx layer ELFs were captured by LD_PRELOAD-interposing AMD FastFlowLM's real `qwen3_npu` runtime. **FastFlowLM ships no Zaya model** (`src/include/models/`: gemma/qwen3/llama/phi4/nanbeige/…; zero zaya hits) — there is no Zaya NPU runtime to capture ELFs from, so the per-ctx Zaya layer ELFs must be authored (generator), not captured.

## Verified building blocks (separate single-dispatch designs today)

1. `attn.xclbin` — QK^T→softmax→PV, N=512 baked, on-core exp-LUT softmax; host bridge `npu_attn_ctx.h` corr ≥ 0.999 (issue #1776 rounds).
2. Fused GU→SiLU→D: decode uses 2-launch GUSILU + m8h2 (6.2 ms/layer); the zero-DMA single-launch cascade family is silicon-verified (Qwen3 real weights EXACT; Zaya decode-geometry artifact built, numerics open — see the cascade issue).
3. Per-ctx layer-ELF + runlist mechanism: merged via #2063 (`RuntimeLayerEngine`, `npu-infer/src/runtime_layer.cpp` forward() pattern).

## Milestones

- M1 (done as foundation): decode phase profile + variant matrix — MoE-bound at 124/156 ms/tok; attention-on-NPU slower; runlist not applicable per-op (measured, logged in M3 doc).
- M2 (in progress): single-launch cascade for Zaya decode geometry (artifact N_D=2048 ROWS=4 built; real-weight numerics OPEN).
- M3: whole-layer kernel authoring — co-place attn + fused MoE in one hw_context; fold host-in-the-middle ops on-device (residual/norm, cca_prep conv_qk/means/L2/RoPE, silu, router); KV → device BO; per-ctx KV-length TXN (mirroring FLM `elf_0001/0003/0004/0005` + lm_head structure).
- M4: `RuntimeLayerEngine`-for-Zaya `forward()` — one `xrt::runlist`/token (copy `runtime_layer.cpp` pattern); numeric parity vs the CPU-float decode + byte-determinism gate; target ≥ 25–40 tok/s.

## Effort

Weeks of AIE2P bring-up (consistent with #1776's original 2–4 wk endgame estimate). Foundation + cascade artifact work is preserved at `~/wt/zaya-m1/engine/npu/docs/M3-ZAYA-WHOLE-LAYER.md` (strixhalo).

Refs: #1776, #2078, #1759/#2063.

**Comment by bong-water-water-bong** [comment]
## M2 CLOSED (cascade numerics done, corr 0.999346) + concrete M3 kickoff scope

**M2 status**: the single-launch cascade for Zaya decode geometry is now silicon-verified at **corr 0.999346** (commit `e3abace8` on `npu/cascade-e2b-preserve`; full chain in #2078). The remaining cascade item is decode-PERF (its correctness-first launch ~12-14 ms vs the tuned 2-launch P1+P2 4.65 ms; the bottleneck is ~1300 serialized per-fill awaits — the batched-fill attempt deadlocks on the runtime BD-reuse; needs the iron rt task_group/BD semantics rework, or porting the launch to the older pipelined-fill API).

**M3 scope — whole-layer per-ctx ELF authoring** (concrete, from the actual decode structure):

The decode's per-token budget: MoE 124 ms (10 layers × 6.2 ms = P1 3.1 + P2 1.5 + pre 1.6 per layer, **host-serialized launches**), attn 10.4 ms (CPU GQA), norm+logits 22 ms. The M3 doc's key measurement: the per-op path cannot be runlist-batched because the HOST sits between kernels (silu/norm/cca_prep/RoPE/quant). M3 = move those on-device + co-place attn+MoE in one hw_context + one runlist/token.

**Work items in order (each silicon-gated):**
1. **Kernel inventory**: enumerate every per-layer host op in `zaya_decode_main`'s forward loop (residual-add, rmsnorm, cca_prep conv_qk/means/L2, RoPE, router, quantize, the MoE GU/silu/D, the layer output) with its current cost + the NPU kernel that would absorb it. (1 session — pure analysis of the existing code + the attn.xclbin/MoE kernels' interfaces.)
2. **The dispatch-latency ceiling**: measure the actual per-launch dispatch overhead in the current decode (the M3 doc estimates 5-9 ms/launch × 4/layer) vs the on-device compute (~3 μs for the M3-sized MACs) — this quantifies the runlist win and gates whether M3's target (25-40 tok/s) is reachable by collapsing launches alone. (1 session — instrument the existing decode.)
3. **On-device residual+norm**: author the RMSNorm + residual-add into a small NPU kernel (the rmsnorm pattern likely exists in the iron operator set) — removes the first host round trips. (2-3 sessions with the iron flow.)
4. **Co-place attn + MoE**: attn.xclbin + the fused-MoE kernels in ONE hw_context with per-token runlist sequencing (the #2063 RuntimeLayerEngine pattern). (the core M3 authoring, multi-week.)
5. **M4**: RuntimeLayerEngine-for-Zaya forward() + byte-determinism gate.

**First concrete deliverable (item 1+2)**: a per-layer kernel/host-op inventory + the measured dispatch-latency breakdown, posted here — that decides whether the whole-layer ELF (authoring) or a cheaper launch-collapse (runlist over the existing per-op kernels with the host ops on CPU) is the right M3 path. The M3 doc's "runlist not applicable per-op" claim was about the EXISTING host-in-the-middle structure — item 2 re-measures it with the current 2-launch fused MoE (P1+P2 only, host silu already on-device) to find the actual remaining collapse opportunity.

Refs: #1776, #2078 (closed), #1759/#2063. M3 log: `engine/npu/docs/M3-ZAYA-WHOLE-LAYER.md`.

**Comment by bong-water-water-bong** [comment]
## M3 kickoff inventory — the per-layer op structure of the Zaya decode

From `zaya_decode.cpp` forward() + the M3 profile (per token):

```
even layer (attn):  h'=(h+hb)*hs; residual += (residual+rb)*rs; cur=rmsnorm(residual)
                    CCA attn on CPU (conv_qk, qk_means, L2, RoPE, GQA)   ~1.0 ms (10 layers)
odd layer (MoE):    cur=rmsnorm(residual)
                    pre: router (CPU) + amax + fused header fold          1.61 ms  (host)
                    P1: GU->SiLU->h2 (final_i8_MOE_GUSILU_zaya)           3.08 ms  (NPU)
                    P2: D-from-h2 (final_i8_MOE_D_zaya_m8h2)              1.49 ms  (NPU)
final: rmsnorm + logits (embed) ~22 ms (int8: 7.6)
```

**Host ops between/around the NPU launches per MoE layer** (the M3 fold targets):
1. residual scale-add (`(residual+rb)*rs`, `h+hb` etc.) — H=2048 elementwise ×2
2. rmsnorm (cur = x/σ·w) — H=2048
3. router (per-token top-k over 16 experts: norm + 2 dense layers) — the 1.61 ms "pre" host block
4. dynamic_ascale (amax pass, H=2048)
5. host_h2_amax_qn_s (the 8M-MAC int8 GU mirror for qn_s — AVX2 ~0.1-0.3 ms)
6. update_fused_header (per-token fold header write)
7. P1→P2 h2 handoff (2 KB DDR round trip, the #1759 fusion already removed the host silu)

**Dispatch-latency reality (correcting the older 5-9 ms/launch claim)**: the CURRENT 2-launch fused kernels run at 3.08 + 1.49 = 4.57 ms/layer NPU-inclusive — the older 5-9 ms dispatch figure predates the #2063/#2053 launch mechanism and does NOT apply to today's P1/P2. So the runlist/whole-layer win is bounded by: (a) the P1→P2 h2 round trip (~0.5-1.2 ms — the cascade's premise, now numerics-closed but not perf-fast), (b) the host "pre" 1.61 ms (router+amax+mirror+header — the biggest host block; folding needs an on-device router + amax, hard), (c) the attn's CPU 1.0 ms.

**M3 implication**: the honest collapse ceiling with today's kernels ≈ removing (a) ~1 ms + partial (b) — NOT the 6.2 → 0.96 ms/layer the Qwen per-ctx ELFs achieved (that came from weights-resident + one launch + no host-in-middle). Reaching ~25-40 tok/s on Zaya needs the whole-layer ELF with the router/amax ALSO on-device (the hard part — the router is a 2-layer dense over 16 experts, amax is a reduction) OR a genuinely co-placed attn+MoE ELF with per-ctx KV. This inventory is posted to #2080 as the M3 kickoff deliverable.

**Comment by bong-water-water-bong** [comment]
**Disposition (triage, 2026-09-05):** OPEN — active feature (#1776 endgame: Zaya whole-layer per-ctx ELF decode, M3/M4). M2 closed today 13:13Z (cascade numerics corr 0.999346); M3 kickoff scoped in the same comment. **Next owner:** cascade/whole-layer lane (strixhalo). Next action: M3 whole-layer kernel authoring + RuntimeLayerEngine-for-Zaya integration per the M3 doc. Keep open as the active feature lane.

**Comment by bong-water-water-bong** [comment]
_Diagnosis + fix plan — issue campaign (goal `mtubusx1-swek0b`, 2026-09-09). Full inventory: `docs/issue-campaign/FIX-LIST.md`._

**Diagnosis: a port/authoring milestone, not plumbing. M1 done, M2 in progress, M3/M4 = the authoring work.**

M1: MoE-bound 124/156 ms/tok; runlist not applicable per-op. M2: single-launch cascade, N_D=2048 ROWS=4 artifact built, numerics open. M3/M4: author whole-layer kernels (co-place attn + fused MoE, fold residual/norm/cca_prep/RoPE/silu/router on-device, KV→device BO) + RuntimeLayerEngine-for-Zaya forward() (one xrt::runlist/token). Authoring (not capture) because FLM ships no Zaya model.

**Fix plan:** (1) close #2113/#2114 (cascade correctness+perf); (2) resolve #2070 (device-VA) only if per-ctx ELFs need fixed addresses; (3) author the whole-layer kernel + RuntimeLayerEngine forward() per M3/M4 spec. **verify:** one runlist/token, numeric parity vs CPU-float, byte-determinism, ≥25–40 tok/s.

—_posted by the issue campaign coordinator_

**Comment by bong-water-water-bong** [comment]
## 35B-A3B is a proven dead end on this path (2026-09-11) — recorded here so it isn't re-litigated

The goal that tried to extend the whole-layer per-ctx ELF runlist from dense Qwen3 to Qwen3.6-35B-A3B (`mtusoiy1-cfdhqr`) ran to completion and stopped on a **negative result**. Two independent causes:

- **The per-ctx 35B layer ELFs produce NaNs** — the region-B int4 generator is closed source (R59). Needs FastFlowLM source, not a wiring change.
- **Cross-xclbin runlist batching is infeasible**: the MoE xclbins carry different ERT `group_id`s, so they cannot be batched into one `xrt::runlist` submit (R98).

Measured on the box: **0.57 tok/s** (CPU MoE, 80 launches/token — 40 QKV + 40 O) and **0.32 tok/s** (NPU fused M=1), with the FFN dominating at 26.6 ms/layer. Both are *below* the ~0.7 tok/s baseline this path started from, and far from the dense-Qwen3 one-submit class the goal targeted.

What that work did produce and is usable:

- 40 MoE layer ELFs (30 linear 24636 w / 630 DDR_PATCH + 10 full-attn 20566 w / 530) + an lm_head ELF, all TXN-verified; they assemble via aiebu. Regions A+B only — experts excluded.
- The MoE weight-BO packers, byte-identical to the reference for layers 0, 1, 6, 20, 30, 38 (QKV/O/router/k/v max|d| = 0.0).
- A `MoERuntimeLayerEngine` (single `xrt::runlist`, layer + lm_head batched) wired into the engine, whose weight layout matches that reference.

Results: `benchmarks/RESULTS-runlist-decode-35b-moe-2026-09-10.md` and `benchmarks/RESULTS-runlist-decode-2026-09-09.md` on branch `goal/runlist-decode-wire`. Also recorded under "Settled negative results" in `research/TRACKING.md` (PR #2306).

**Conclusion for this issue:** the 35B is not a "finish the wiring" item — it is blocked on a region-B int4 generator we do not have. Re-opening it needs that first, or the MoE path needs a different decomposition than whole-layer per-ctx ELFs.

**Comment by bong-water-water-bong** [comment]
Re-checked as of 2026-09-22: `~/wt/zaya-m1/engine/npu/docs/M3-ZAYA-WHOLE-LAYER.md` on strixhalo unchanged since 2026-09-05 — no progress on M3 (whole-layer kernel authoring) or M4 (RuntimeLayerEngine-for-Zaya). This issue's own effort estimate is "weeks of AIE2P bring-up," consistent with the original #1776 2-4wk endgame estimate.

Status unchanged: PARKED, scheduled major authoring work — not something a triage pass advances.
